### PR TITLE
minor: Avoid per cell evaluation in Coalesce, use zip in CaseWhen

### PR DESCRIPTION
**Pull request by yjshen**
# Which issue does this PR close?

Closes #.

 # Rationale for this change
- [Tests](https://github.com/blaze-init/spark-blaze-extension/runs/5847348229?check_suite_focus=true) with Coalesce in Blaze seem to run forever because we are iterating and calculating per cell. 
- `If_then_else` in `case.rs` could be substitute by `arrow::compute::kernels::zip`

# What changes are included in this PR?
- Rework a little bit on Coalesce evaluation.
- Use `zip` in `CaseWhen` evaluation.

# Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/case.rs [hunk 1]
```diff
@@ -19,7 +19,8 @@ use std::{any::Any, sync::Arc};
 
 use crate::expressions::try_cast;
 use crate::PhysicalExpr;
-use arrow::array::{self, *};
+use arrow::array::*;
+use arrow::compute::kernels::zip::zip;
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/case.rs:23]: Today I Learned: https://docs.rs/arrow/11.1.0/arrow/compute/kernels/zip/fn.zip.html ✅

File: datafusion/physical-expr/src/conditional_expressions.rs [hunk 2]
```diff
@@ -35,36 +33,51 @@ pub fn coalesce(args: &[ColumnarValue]) -> Result<ColumnarValue> {
...
-                ColumnarValue::Array(array_ref) => {
-                    let curr_null_mask = compute::is_null(res.as_ref())?;
-                    let arr_not_null_mask = compute::is_not_null(array_ref)?;
-                    let bool_mask = compute::and(&curr_null_mask, &arr_not_null_mask)?;
-                    res = zip(&bool_mask, array_ref, &res)?;
+        for arg in args {
+            match arg {
+                ColumnarValue::Array(ref array) => {
+                    let to_apply = and(&remainder, &is_not_null(array.as_ref())?)?;
+                    current_value = zip(&to_apply, array, current_value.as_ref())?;
```
Comment (alamb) [inline datafusion/physical-expr/src/conditional_expressions.rs:51]: 👍

File: datafusion/physical-expr/src/conditional_expressions.rs [hunk 3]
```diff
@@ -35,36 +33,51 @@ pub fn coalesce(args: &[ColumnarValue]) -> Result<ColumnarValue> {
...
-                        res =
-                            zip(&bool_arr, scalar.to_array_of_size(size).as_ref(), &res)?;
+                ColumnarValue::Scalar(value) => {
+                    if value.is_null() {
                         continue;
+                    } else {
+                        let last_value = value.to_array_of_size(size);
+                        current_value =
+                            zip(&remainder, &last_value, current_value.as_ref())?;
+                        break;
```
Comment (alamb) [inline datafusion/physical-expr/src/conditional_expressions.rs:61]: 👍  early return on constant

File: datafusion/physical-expr/src/conditional_expressions.rs [hunk 4]
```diff
@@ -35,36 +33,51 @@ pub fn coalesce(args: &[ColumnarValue]) -> Result<ColumnarValue> {
...
                     }
                 }
             }
+            if remainder.iter().all(|x| x == Some(false)) {
```
Comment (alamb) [inline datafusion/physical-expr/src/conditional_expressions.rs:65]: There are fancier / more optimized ways in the guts of the arrow-rs implementation for testing if a bitmask is entirely false, but maybe we can expose them if more optimization is needed

**Review by alamb — APPROVED** [review]
Nice work @yjshen

File: datafusion/physical-expr/src/conditional_expressions.rs [hunk 5]
```diff
@@ -35,36 +33,51 @@ pub fn coalesce(args: &[ColumnarValue]) -> Result<ColumnarValue> {
...
 
-    let size = match args[0] {
-        ColumnarValue::Array(ref a) => a.len(),
-        ColumnarValue::Scalar(ref _s) => 1,
-    };
-    let mut res = new_null_array(&args[0].data_type(), size);
+    let return_type = args[0].data_type();
+    let mut return_array = args.iter().filter_map(|x| match x {
+        ColumnarValue::Array(array) => Some(array.len()),
+        _ => None,
```
Comment (liukun4515) [inline datafusion/physical-expr/src/conditional_expressions.rs:39]: why not use the same logic by returning `Some(1)`?

Comment (yjshen) [reply datafusion/physical-expr/src/conditional_expressions.rs:39]: return_array is used to decide whether to return a scalar or a vector, so if we see no array sizes and get None from `return_array.next()` we know all params are scalar and will enter branch in line-70 and do fast scalar value coalesce.

**Review by liukun4515 — APPROVED** [review]

**Comment by yjshen** [comment]
Thanks @alamb @liukun4515

File: datafusion/physical-expr/src/conditional_expressions.rs (hunk 4 above)
Comment (Dandandan) [reply datafusion/physical-expr/src/conditional_expressions.rs:65]: There also seem to be some opportunities to optimize `zip` in arrow-rs for all-true or all-false.

File: datafusion/physical-expr/src/expressions/case.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/case.rs:23]: it certainly isn't the same as `zip` in Rust's iterators -- so I agree it is confusing

**Pull request merged**